---
title: "C42B cells treatment comparison"
author: "Marion Hardy"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true 
    theme: spacelab 
    highlight: monochrome
editor_options: 
  chunk_output_type: console
  markdown: 
    wrap: 72
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE, cache = TRUE, echo = FALSE, warning = F, cache.lazy = F)
knitr::opts_chunk$set(fig.width=6, fig.height=7) 

library(RColorBrewer)
library(viridis)
library(tidyverse)
library(DESeq2)
library(cowplot)
library(dendextend)
library(usedist)
library(biomaRt)
library(ggrepel)
library(pheatmap)
library(ComplexHeatmap)
library(xlsx)
library(msigdbr)
library(clusterProfiler)
library(kableExtra)
library(gridExtra)
library(AMR)

ensembl_to_geneName <- readRDS("./data/hsapiens_annotations_230510.rds") %>% 
  filter(!duplicated(gene, drop = F))

dir = getwd()

```

# Introduction

**This report updated "Abi Tala" to "Abi Ola" double treatment.**

Day 5 bulkRNAseq analysis of C42B cells treated with

-   Ola

-   Enz

-   Tal

-   Abi + Ola

-   Enz + Tala

This report will present a broad overview of how similar/different
samples are in terms of gene expression. The objective is to help AK
narrow down which gene sets are of interest and follow this with a gsea
analysis and a closer look at the function of these identified gene
groups.This is an ongoing effort to characterize their DTP/DTEP platform
and possibly understand the mechanisms underlying persistence
acquisition and maintenance.

Each sample comparison shown in this report will have additional files
generated:

-   res_tbl.csv = all differentially expressed genes, annotated (no
    filtering)

-   Sign_genes.xlsx = excel doc with all significantly up- and
    down-regulated genes (log2Fc \>= 1), annotated and sorted into all
    genes, only up and only down on different sheets.

-   volcano plot of all genes (html format)

NB: Data quality has been checked within the previous script "DESeq2.R",
everything looked good.

```{r}

dds = readRDS("./data_output/C42B_DESeq_comp_to_DMSO.Rds")
ddsao = readRDS("./data_output/C42B_DESeq_comp_to_AbiOla.Rds")
ddseo = readRDS("./data_output/C42B_DESeq_comp_to_EnzTala.Rds")

```

```{r}

# MsigDb

hsa_GOBP_sets = msigdbr(
  species = "Homo sapiens", 
  category = "C5",
  subcategory = "GO:BP")

hsa_hallmark_sets = msigdbr(
  species = "Homo sapiens", 
  category = "H")

hsa_reactome_sets = msigdbr(
  species = "Homo sapiens", 
  category = "C2",
  subcategory = "CP:REACTOME") # for reactome collection

set.seed(054057) # for reproducibility

```

# PCA

```{r, fig.height=5, fig.width=6}
rld = vst(dds)

plotPCA(rld,intgroup="Treatment") + 
  geom_label_repel(aes(label = Treatment), 
                   size = 3, box.padding = 0.6,
                   max.overlaps = Inf)+
  xlim(c(-25,20))+
  ylim(c(-15, 15))+
  theme_bw()+
  labs(title = 'PCA per treatment and group')

ggsave(plot = last_plot(),"./figures/PCA1.svg", dpi = 300,
       height = 4.5, width = 5.5)

```

# Extract the weights per genes

Table for top 200 genes (absolute value) to weigh on PC1 and PC2.

```{r, fig.height=5, fig.width=6, echo=FALSE}

data = assay(dds)
forlabel = as.data.frame(rownames(data))
colnames(forlabel) = "ensembl"
forlabel = left_join(forlabel, ensembl_to_geneName)
rownames(data) = forlabel$gene

pca = prcomp(data)
loadings = as.data.frame(pca$x)
loadings$gene = rownames(loadings)

top_genes = loadings %>% 
  select(gene, PC1, PC2) %>%
  pivot_longer(matches("PC"), names_to = "PC", values_to = "loading") %>% 
  group_by(PC) %>% 
  arrange(desc(abs(loading))) %>% 
  slice(1:20) %>% # take the 20 top rows
  pull(gene) %>% 
  unique()

loadings %>% 
  select(gene, PC1, PC2) %>%
  pivot_longer(matches("PC"), names_to = "PC", values_to = "loading") %>% 
  group_by(PC) %>% 
  arrange(desc(abs(loading))) %>% 
  slice(1:200) %>% 
  as_tibble() %>%
  knitr::kable()%>% 
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>% 
  scroll_box(width = "1000px", height = "500px")

write.csv2(loadings,"./data_output/loadings.csv")

```

```{r, fig.height=5, fig.width=6, echo=FALSE}

# plot the biplot

loadings %>% 
  filter(gene %in% top_genes) %>% 
  ggplot() +
  geom_segment(aes(x = 0, y = 0, xend = PC1, yend = PC2), 
               arrow = arrow(length = unit(0.1, "in")),
               color = "firebrick3") +
  geom_text(aes(x = PC1, y = PC2, label = gene),
            nudge_y = 10000, nudge_x = -50000,
            size = 3)+
  theme_bw()+
  xlab("PC1 loadings")+
  ylab("PC2 loadings")+
  labs(title = 'Biplot of gene loadings')

ggsave(plot = last_plot(),"./figures/PCA1_weights.svg", dpi = 300,
       height = 4.5, width = 5.5)


```

# Euclidian distances

```{r, fig.height=5, fig.width=6}
sampleDists = dist(t(assay(rld)))
sampleDistMatrix = as.matrix(sampleDists)
rownames(sampleDistMatrix) = paste(rld$Treatment, sep="-")
colnames(sampleDistMatrix) = paste(rld$Treatment, sep="-")
colors = viridis(20)

pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)
# figure exported as svg manually
```

# Dendogram

```{r, fig.height=4, fig.width=7}

sampleDists_named = sampleDists
sampleDists_named = dist_setNames(sampleDists_named, as.vector(colData(dds)[,4]))

hc = hclust(sampleDists_named)
hc %>% 
  as.dendrogram() %>% 
  set("labels_col", c(1:7),k=8) %>% 
  plot()
# figure exported as svg manually

```

# Every treatment compared to DMSO

## Volcano plots

```{r, fig.show="hold", out.width="33%"}

for (i in 2:7){
    cond = resultsNames(dds)[i]
    name = gsub("Treatment_","",resultsNames(dds)[i])
    
    res <- results(dds, name = cond)
    res_tbl <- as_tibble(res, rownames="ensembl")
    res_tbl <- res_tbl %>%
    left_join(ensembl_to_geneName) %>%
    arrange(padj) 
    dir.create(paste0(dir,'/data_output/',name))
    dir.create(paste0(dir,'/figures/',name))
    write.csv(res_tbl, paste0("./data_output/",name,"/res_tbl.csv"), row.names = T)
    
    overexpr <- res_tbl %>% filter(padj<=0.05 & log2FoldChange>=0.5) 
    underexpr <- res_tbl %>% filter(padj<=0.05 & log2FoldChange<=-0.5)
    signif <- full_join(overexpr, underexpr)
  
    write.xlsx(overexpr, file= paste0("./data_output/",name,"/Sign_genes.xlsx"), sheetName="UP")
    write.xlsx(underexpr, file=paste0("./data_output/",name,"/Sign_genes.xlsx"), sheetName="DOWN", append=TRUE)
    write.xlsx(signif, file=paste0("./data_output/",name,"/Sign_genes.xlsx"), sheetName="SIGNIF", append=TRUE)
    
    p1 =
  res_tbl %>%
    filter(!is.na(padj)) %>%
    ggplot(aes(x = log2FoldChange, y = -log10(padj),
               color = padj < 0.05 & abs(log2FoldChange) > 0.5, 
               label = ifelse(padj<0.05&log2FoldChange>=0.5|
                                padj<0.05&log2FoldChange<=-0.5,as.character(gene),''))) +
    scale_colour_manual(values = c("gray", "firebrick3")) +
    labs(col="Significantly expressed")+
    geom_point(size = 0.5) +
    geom_hline(yintercept = -log10(0.05)) +
    geom_vline(xintercept = 0.5) +
    geom_vline(xintercept = -0.5)+
    geom_text_repel(max.overlaps = 15,
                    box.padding = 0.25,
                    segment.color = 'grey50',
                    fontface = "italic")+
    labs(title = name)+
    theme_bw()
  
  print(p1)
  
    ggsave(paste0("./figures/",name,"/volcanoplot.svg"), last_plot(), 
           dpi= 500, width = 12, height = 8)
  
}

```

# Abi and Ola vs Tala+Abi

## Volcano plots

```{r, fig.show="hold", out.width="50%"}

for (i in c(3,6)){
    cond = resultsNames(ddsao)[i]
    name = gsub("Treatment_","",resultsNames(ddsao)[i])
    
    res <- results(ddsao, name = cond)
    res_tbl <- as_tibble(res, rownames="ensembl")
    res_tbl <- res_tbl %>%
    left_join(ensembl_to_geneName) %>%
    arrange(padj) 
    dir.create(paste0(dir,'/data_output/',name))
    dir.create(paste0(dir,'/figures/',name))
    write.csv(res_tbl, paste0("./data_output/",name,"/res_tbl.csv"), row.names = T)
    overexpr <- res_tbl %>% filter(padj<=0.05 & log2FoldChange>=0.5) 
    underexpr <- res_tbl %>% filter(padj<=0.05 & log2FoldChange<=-0.5)
    signif <- full_join(overexpr, underexpr)
  
    write.xlsx(overexpr, file= paste0("./data_output/",name,"/Sign_genes.xlsx"), sheetName="UP")
    write.xlsx(underexpr, file=paste0("./data_output/",name,"/Sign_genes.xlsx"), sheetName="DOWN", append=TRUE)
    write.xlsx(signif, file=paste0("./data_output/",name,"/Sign_genes.xlsx"), sheetName="SIGNIF", append=TRUE)
    
    p1 =
  res_tbl %>%
    filter(!is.na(padj)) %>%
    ggplot(aes(x = log2FoldChange, y = -log10(padj),
               color = padj < 0.05 & abs(log2FoldChange) > 0.5, 
               label = ifelse(padj<0.05&log2FoldChange>=0.5|
                                padj<0.05&log2FoldChange<=-0.5,as.character(gene),''))) +
    scale_colour_manual(values = c("gray", "firebrick3")) +
    labs(col="Significantly expressed")+
    geom_point(size = 0.5) +
    geom_hline(yintercept = -log10(0.05)) +
    geom_vline(xintercept = 0.5) +
    geom_vline(xintercept = -0.5)+
    geom_text_repel(max.overlaps = 15,
                    box.padding = 0.25,
                    segment.color = 'grey50',
                    fontface = "italic")+
    labs(title = name)+
    theme_bw()
  
  print(p1)
  
    ggsave(paste0("./figures/",name,"/volcanoplot.svg"), last_plot(), 
           dpi= 500, width = 12, height = 8)
  
    
}


```

# Enz and Ola vs Tala+Abi

## Volcano plots

```{r, fig.show="hold", out.width="50%"}

for (i in c(5,6)){
    cond = resultsNames(ddseo)[i]
    name = gsub("Treatment_","",resultsNames(ddseo)[i])
    
    res <- results(ddseo, name = cond)
    res_tbl <- as_tibble(res, rownames="ensembl")
    res_tbl <- res_tbl %>%
    left_join(ensembl_to_geneName) %>%
    arrange(padj) 
    dir.create(paste0(dir,'/data_output/',name))
    dir.create(paste0(dir,'/figures/',name))
    write.csv(res_tbl, paste0("./data_output/",name,"/res_tbl.csv"), row.names = T)
    overexpr <- res_tbl %>% filter(padj<=0.05 & log2FoldChange>=0.5) 
    underexpr <- res_tbl %>% filter(padj<=0.05 & log2FoldChange<=-0.5)
    signif <- full_join(overexpr, underexpr)
  
    write.xlsx(overexpr, file= paste0("./data_output/",name,"/Sign_genes.xlsx"), sheetName="UP")
    write.xlsx(underexpr, file=paste0("./data_output/",name,"/Sign_genes.xlsx"), sheetName="DOWN", append=TRUE)
    write.xlsx(signif, file=paste0("./data_output/",name,"/Sign_genes.xlsx"), sheetName="SIGNIF", append=TRUE)
    
    p1 =
  res_tbl %>%
    filter(!is.na(padj)) %>%
    ggplot(aes(x = log2FoldChange, y = -log10(padj),
               color = padj < 0.05 & abs(log2FoldChange) > 0.5, 
               label = ifelse(padj<0.05&log2FoldChange>=0.5|
                                padj<0.05&log2FoldChange<=-0.5,as.character(gene),''))) +
    scale_colour_manual(values = c("gray", "firebrick3")) +
    labs(col="Significantly expressed")+
    geom_point(size = 0.5) +
    geom_hline(yintercept = -log10(0.05)) +
    geom_vline(xintercept = 0.5) +
    geom_vline(xintercept = -0.5)+
    geom_text_repel(max.overlaps = 15,
                    box.padding = 0.25,
                    segment.color = 'grey50',
                    fontface = "italic")+
    labs(title = name)+
    theme_bw()
  
  print(p1)
  
    ggsave(paste0("./figures/",name,"/volcanoplot.svg"), last_plot(), 
           dpi= 500, width = 12, height = 8)
  
}


```

# Venn diagram on steroids (UpsetPlot)

Conditions we want to compare:

-   Treatment_Abi_5_vs_DMSO

-   Treatment_Enz_20_vs_DMSO

-   Treatment_Ola_1_vs_DMSO

-   Treatment_Tala_10_vs_DMSO

-   Treatment_Abi_Ola_vs_DMSO

-   Treatment_Enz_Tala_vs_DMSO

-   Treatment_Ola_1_vs_Abi_Ola

-   Treatment_Abi_5_vs_Abi_Ola

-   Treatment_Enz_20_vs_Enz_Tala

-   Treatment_Ola_1_vs_Enz_Tala

I extracted all significantly differentially expressed genes (DEGs) for
each condition (\|log2Fc\| \>= 1 and pval \< 0.05).

**Set size** = how many DEGs there are per sample

**Intersection size** = How many genes are shared between the x samples
tied together with the bar + nodes

So if you sum up all the intersection size with a dot in a sample, you
get the set size of that sample. It's essentially a venn diagram for
when you have \>3 samples you want to compare.

```{r, fig.height=5, fig.width=10}

# load in the data to be formatted for an upset plot
# keep each gene name that has a signif pval and |log2fc| > 0.5
# use ensembl to encompass unnamed transcripts

folder = c("Abi_5_vs_DMSO","Enz_20_vs_DMSO","Ola_1_vs_DMSO","Tala_10_vs_DMSO",
           "Abi_Ola_vs_DMSO","Enz_Tala_vs_DMSO","Ola_1_vs_Abi_Ola",
           "Abi_5_vs_Abi_Ola","Enz_20_vs_Enz_Tala","Ola_1_vs_Enz_Tala")

for(i in 1:length(folder)){
  res = read.csv(paste0("./data_output/",folder[i],"/res_tbl.csv"))
  genes = res %>% 
    select(ensembl, log2FoldChange, padj) %>% 
    filter(abs(log2FoldChange)>= 0.5 & padj < 0.05) %>% 
    select(ensembl)
  genes$temp = 1
  colnames(genes)[2] = folder[i]
  assign(folder[i], genes)
}

datalist_dmso = list(Abi_5_vs_DMSO, Enz_20_vs_DMSO, Ola_1_vs_DMSO, Tala_10_vs_DMSO,Abi_Ola_vs_DMSO , Enz_Tala_vs_DMSO )

datalist_bith = list(Ola_1_vs_Abi_Ola, 
             Abi_5_vs_Abi_Ola, Enz_20_vs_Enz_Tala, Ola_1_vs_Enz_Tala)

big_data_dmso =
  datalist_dmso %>% purrr::reduce(full_join, by='ensembl')%>% 
  mutate(across(where(is.numeric), ~ replace_na(.x, 0))) 

big_data_bith =
  datalist_bith %>% purrr::reduce(full_join, by='ensembl')%>% 
  mutate(across(where(is.numeric), ~ replace_na(.x, 0))) 

```

```{r, fig.height=4, fig.width=10}

m1 = make_comb_mat(big_data_dmso)

UpSet(m1, top_annotation = upset_top_annotation(m1, add_numbers = TRUE),
    right_annotation = upset_right_annotation(m1, add_numbers = TRUE),
    comb_order = order(-comb_size(m1)), 
    comb_col = c("#ffbe4f", "#e8702a", "#6bd2db","#7e073d","#0c457d","black")[comb_degree(m1)])
grid.text("All treatments vs DMSO: Gene overlap",x = 0.5, y=0.95, gp=gpar(fontsize=15))

```

```{r, fig.height=4, fig.width=10}

m2 = make_comb_mat(big_data_bith)

UpSet(m2, top_annotation = upset_top_annotation(m2, add_numbers = TRUE),
    right_annotation = upset_right_annotation(m2, add_numbers = TRUE),
    comb_order = order(-comb_size(m2)), 
    comb_col = c("#ffbe4f", "#e8702a", "#6bd2db","#7e073d")[comb_degree(m2)])
grid.text("Mono- vs bi-therapies: Gene overlap",x = 0.5, y=0.95, gp=gpar(fontsize=15))

```
